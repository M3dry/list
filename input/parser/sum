(+ 4 1)

-- args: -i -p -f -e
-- expected stdout:
-- Ok(
--     Call(
--         Call {
--             func: Right(
--                 "+",
--             ),
--             params: [
--                 Literal(
--                     Int(
--                         false,
--                         4,
--                     ),
--                 ),
--                 Literal(
--                     Int(
--                         false,
--                         1,
--                     ),
--                 ),
--             ],
--         },
--     ),
-- )

