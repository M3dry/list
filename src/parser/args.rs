use crate::tokenizer::{Keywords, Token};

use super::{
    error,
    r#type::{Generic, Type},
    Parser, ParserError, ParserErrorStack,
};

#[derive(Debug)]
pub(crate) enum Arg {
    Generic(Generic),
    Named(String, Type),
    Simple(String),
    SelfA(Type),
}

impl TryFrom<&mut Parser> for Arg {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let next = value.pop_front_err("Arg")?;
        let name = match next {
            Token::Identifier(iden) if &iden == "self" => return Ok(Arg::SelfA(Type::SelfA)),
            w @ Token::Ref => {
                value.tokens.push_front(w);
                let ret = error!(Type::try_from(&mut *value), "Arg")?;

                match &ret {
                    Type::Ref(_, t) | Type::RefMut(_, t) if matches!(**t, Type::SelfA) => {
                        return Ok(Self::SelfA(ret))
                    }
                    token => return Err(error!("Arg", format!("Expected self, got {token:#?}"))),
                }
            }
            w @ Token::Char(':') => {
                value.tokens.push_front(w);
                return Ok(Arg::Generic(error!(Generic::try_from(&mut *value), "Arg")?));
            }
            Token::Identifier(iden) => iden,
            _ => {
                return Err(error!(
                    "Arg name",
                    format!("Expected identifier, got {next:#?}"),
                ));
            }
        };

        let next = value.first();
        if next != Some(&Token::Keyword(Keywords::LeftArrow)) {
            return Ok(Arg::Simple(name));
        }
        value.pop_front();

        let arg_type = error!(Type::try_from(&mut *value), "Arg type")?;

        Ok(Arg::Named(name, arg_type))
    }
}

#[derive(Debug)]
pub struct ArgsTyped {
    lifetimes: Vec<String>,
    generics: Vec<Generic>,
    selft: Option<Type>,
    args: Vec<(String, Type)>,
}

impl TryFrom<&mut Parser> for ArgsTyped {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let mut args = vec![];
        let mut generics = vec![];
        let mut lifetimes = vec![];

        let next = value.pop_front_err("ArgsTyped")?;
        if next != Token::ParenOpen {
            return Err(error!(
                "ArgsTyped",
                format!("Expected ParenOpen, got {next:#?}"),
            ));
        }

        loop {
            let peek = value
                .first()
                .ok_or(error!("ArgsTyped", format!("Expected more tokens")))?;

            match peek {
                Token::Char(':') => {
                    generics.push(error!(Generic::try_from(&mut *value), "ArgsTyped")?)
                }
                Token::BackTick => {
                    value.pop_front();
                    match value.pop_front_err("ArgsTyped")? {
                        Token::Identifier(iden) => lifetimes.push(iden),
                        token => {
                            return Err(error!(
                                "ArgsTyped",
                                format!("Expected identifier, got {token:#?}")
                            ))
                        }
                    }
                }
                _ => break,
            }
        }

        let mut selft = None;

        loop {
            if value.first() == Some(&Token::ParenClose) {
                value.pop_front();
                break;
            }
            let arg = error!(Arg::try_from(&mut *value), "ArgsTyped")?;
            match arg {
                Arg::Generic(_) => return Err(error!("ArgsTyped", format!("Expected a named arg, got a generic, those should be defined before named args"))),
                Arg::Simple(_) => {
                    return Err(error!(
                        "ArgsTyped",
                        format!("Expected named arg, got a simple arg"),
                    ))
                }
                Arg::SelfA(selft2) => {
                    selft = Some(selft2)
                },
                Arg::Named(name, arg_type) => args.push((name, arg_type)),
            }
        }

        Ok(ArgsTyped {
            lifetimes,
            selft,
            generics,
            args,
        })
    }
}

impl ToString for ArgsTyped {
    fn to_string(&self) -> String {
        format!(
            "{}({}{})",
            if !self.generics.is_empty() {
                format!(
                    "<{}>",
                    &self.generics.iter().fold(String::new(), |str, generic| {
                        format!("{str}, {}", generic.to_string())
                    })[2..]
                )
            } else {
                format!("")
            },
            if let Some(selft) = &self.selft {
                format!(
                    "{}{}",
                    selft.to_string(),
                    if self.args.is_empty() {
                        format!("")
                    } else {
                        format!(", ")
                    }
                )
            } else {
                format!("")
            },
            if !self.args.is_empty() {
                (&self.args.iter().fold(String::new(), |str, arg| {
                    format!("{str}, {}: {}", arg.0, arg.1.to_string())
                })[2..])
                    .to_string()
            } else {
                format!("")
            }
        )
    }
}

#[derive(Debug)]
pub struct Args(Vec<String>);

impl TryFrom<&mut Parser> for Args {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let next = value.pop_front_err("Args")?;
        if next != Token::ParenOpen {
            return Err(error!("Args", format!("Expected ParenOpen, got {next:#?}"),));
        }

        let mut args = vec![];
        loop {
            if value.first() == Some(&Token::ParenClose) {
                value.pop_front();
                break;
            }

            let arg = error!(Arg::try_from(&mut *value), "Args")?;
            match arg {
                Arg::Simple(name) => args.push(name),
                _ => return Err(error!("Args arg", format!("Expected just simple args"),)),
            }
        }

        Ok(Self(args))
    }
}

impl ToString for Args {
    fn to_string(&self) -> String {
        format!(
            "|{}|",
            &self
                .0
                .iter()
                .fold(String::new(), |str, arg| { format!("{str}, {arg}") })[2..]
        )
    }
}
