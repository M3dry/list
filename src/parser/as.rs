use crate::tokenizer::{Keywords, Token};

use super::{
    error,
    exp::Exp,
    r#type::Type,
    Parser, ParserError, ParserErrorStack,
};

#[derive(Debug)]
pub struct As {
    exp: Exp,
    to: Type,
}
impl TryFrom<&mut Parser> for As {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let next = value.pop_front_err("As")?;
        if next != Token::ParenOpen {
            return Err(error!("As", format!("Expected parenOpen, got {next:#?}")));
        }
        let next = value.pop_front_err("As")?;
        if next != Token::Keyword(Keywords::As) {
            return Err(error!("As", format!("Expected as, got {next:#?}")));
        }

        let exp = error!(Exp::try_from(&mut *value), "Exp")?;
        let to = error!(Type::try_from(&mut *value), "Exp")?;

        let next = value.pop_front_err("As")?;
        if next != Token::ParenClose {
            return Err(error!("As", format!("Expected parenClose, got {next:#?}")));
        }

        Ok(Self { exp, to })
    }
}

impl ToString for As {
    fn to_string(&self) -> String {
        format!("({} as {})", self.exp.to_string(), self.to.to_string())
    }
}
