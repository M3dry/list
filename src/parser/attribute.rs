use crate::tokenizer::Token;

use super::{error, exp::Exp, Parser, ParserError, ParserErrorStack};

#[derive(Debug)]
pub enum Attribute {
    Inner(AttributeOps),
    Outer(AttributeOps),
}

impl TryFrom<&mut Parser> for Attribute {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let _ = error!("Attribute", value.pop_front(), [Token::Char('#')])?;
        let next = error!(
            "Attribute",
            value.pop_front(),
            [Token::Char('!'), Token::BracketOpen]
        )?;
        let inner = if next == Token::Char('!') {
            let _ = error!("Attribute", value.pop_front(), [Token::BracketOpen])?;
            true
        } else {
            false
        };

        let attr_ops = error!(AttributeOps::try_from(&mut *value), "Attribute")?;
        let _ = error!("Attribute", value.pop_front(), [Token::BracketClose])?;

        Ok(if inner {
            Self::Inner(attr_ops)
        } else {
            Self::Outer(attr_ops)
        })
    }
}

impl ToString for Attribute {
    fn to_string(&self) -> String {
        match self {
            Attribute::Inner(attr_ops) => format!("#![{}]", attr_ops.to_string()),
            Attribute::Outer(attr_ops) => format!("#[{}]", attr_ops.to_string()),
        }
    }
}

#[derive(Debug)]
pub enum AttributeOps {
    Command(String, Vec<AttributeOps>),
    Assignment(String, Exp),
    Identifier(String),
}

impl TryFrom<&mut Parser> for AttributeOps {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let next = error!(
            "AttributeOps",
            value.pop_front(),
            [Token::ParenOpen, Token::Identifier(_)]
        )?;
        Ok(match next {
            Token::ParenOpen => match error!(
                "AttributeOps",
                value.pop_front(),
                [Token::Identifier(_), Token::Char('=')]
            )? {
                Token::Identifier(iden) => {
                    let mut args = vec![];

                    loop {
                        let peek = value.first_err("AttributeOps")?;

                        if peek == &Token::ParenClose {
                            value.pop_front();
                            break Self::Command(iden, args);
                        }

                        args.push(error!(AttributeOps::try_from(&mut *value), "AttributeOps")?)
                    }
                }
                Token::Char('=') => {
                    let iden = error!("AttributeOps", value);
                    let exp = error!(Exp::try_from(&mut *value), "AttributeOps")?;
                    let _ = error!("AttributeOps", value.pop_front(), [Token::ParenClose])?;

                    Self::Assignment(iden, exp)
                }
                _ => unreachable!(),
            },
            Token::Identifier(iden) => Self::Identifier(iden),
            _ => unreachable!(),
        })
    }
}

impl ToString for AttributeOps {
    fn to_string(&self) -> String {
        match self {
            Self::Command(name, args) => format!(
                "{name}({})",
                &if args.is_empty() {
                    format!(", ")
                } else {
                    args.iter().fold(String::new(), |str, arg| {
                        format!("{str}, {}", arg.to_string())
                    })
                }[2..]
            ),
            Self::Assignment(var, exp) => format!("{var} = {}", exp.to_string()),
            Self::Identifier(var) => var.to_string(),
        }
    }
}
