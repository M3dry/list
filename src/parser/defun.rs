use crate::tokenizer::{Keywords, Token};

use super::{
    args::ArgsTyped, error, exp::Exp, r#type::Type, Parser, ParserError, ParserErrorStack,
};

#[derive(Debug)]
pub struct Defun {
    scope: Scope,
    name: String,
    args: ArgsTyped,
    return_type: Type,
    body: Exp,
}

impl TryFrom<&mut Parser> for Defun {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let next = value.pop_front_err("Defun")?;
        if next != Token::ParenOpen {
            return Err(error!(
                "Defun",
                format!("Expected ParenOpen, got {next:#?}"),
            ));
        }

        let scope = Scope::try_from(&mut *value).unwrap();

        let next = value.pop_front_err("Defun")?;
        if next != Token::Keyword(Keywords::Defun) {
            return Err(error!(
                "Defun",
                format!("Expected Defun keyword, got {next:#?}"),
            ));
        }

        let next = value.pop_front_err("Defun")?;
        let name = if let Token::Identifier(iden) = next {
            iden
        } else {
            return Err(error!(
                "Defun name",
                format!("Expected Identifier, got {next:#?}"),
            ));
        };

        let args = error!(ArgsTyped::try_from(&mut *value), "Defun")?;

        let next = value.pop_front_err("Defun")?;
        if next != Token::Keyword(Keywords::Arrow) {
            return Err(error!(
                "Defun arrow",
                format!("Expected Arrow keyword, got {next:#?}"),
            ));
        }

        let return_type = error!(Type::try_from(&mut *value), "Defun")?;
        let body = error!(Exp::try_from(&mut *value), "Defun")?;

        Ok(Defun {
            scope,
            name,
            args,
            return_type,
            body,
        })
    }
}

impl ToString for Defun {
    fn to_string(&self) -> String {
        format!(
            "{}fn {}{} -> {} {{{}}}",
            self.scope.to_string(),
            self.name,
            self.args.to_string(),
            self.return_type.to_string(),
            self.body.to_string()
        )
    }
}

#[derive(Debug)]
enum Scope {
    File,
    Crate,
    Full,
}

impl TryFrom<&mut Parser> for Scope {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let next = value
            .first()
            .ok_or(error!("Scope", format!("Expected more tokens")))?;
        match next {
            Token::Identifier(iden) => match &iden[..] {
                "crate" => {
                    value.pop_front();
                    Ok(Scope::Crate)
                }
                "pub" => {
                    value.pop_front();
                    Ok(Scope::Full)
                }
                iden => Err(error!("Scope iden", format!("Expected pub, got {iden:#?}"),)),
            },
            _ => Ok(Scope::File),
        }
    }
}

impl ToString for Scope {
    fn to_string(&self) -> String {
        match self {
            Scope::File => "",
            Scope::Crate => "pub(crate) ",
            Scope::Full => "pub ",
        }
        .to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::snapshot;

    snapshot!(test_defun, Defun::try_from, "defun.lt");
    snapshot!(test_defun_rust, Defun::try_from, "defun.lt", rust);
}
