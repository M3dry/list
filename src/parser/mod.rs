pub mod args;
pub mod r#as;
pub mod attribute;
pub mod defun;
pub mod r#do;
pub mod r#enum;
pub mod exp;
pub mod file;
pub mod r#if;
pub mod r#impl;
pub mod lambda;
pub mod r#let;
pub mod r#match;
pub mod range;
pub mod r#struct;
pub mod r#type;
pub mod r#use;
pub mod r#trait;
pub mod module;

use std::{collections::VecDeque, path::PathBuf};

use std::ops::Index;

use crate::tokenizer::{Token, Tokens};

macro_rules! error {
    ($error:expr, $name:literal) => {
        $error.map_err(|mut err| {
            err.stack.push(ParserErrorStack {
                name: $name,
                file: file!(),
                location: (line!(), column!()),
            });
            err
        })
    };
    ($initial:expr, $err:expr$(,)?) => {
        ParserError {
            stack: vec![ParserErrorStack {
                name: $initial,
                file: file!(),
                location: (line!(), column!()),
            }],
            err: $err,
        }
    };
}

#[allow(unused_macros)]
macro_rules! snapshot {
    ($name:tt, $func:expr, $path:tt) => {
        snapshot!(
            $name,
            |parser| format!("{:#?}", $func(parser)),
            $path,
            "../../testdata/parser/"
        );
    };
    ($name:tt, $func:expr, $path:tt, rust) => {
        snapshot!(
            $name,
            |parser| match $func(parser) {
                Ok(res) => res.to_string(),
                Err(err) => format!("{err:#?}"),
            },
            $path,
            "../../testdata/rust/"
        );
    };
    ($name:tt, $func:expr, $path:tt, $out:literal) => {
        #[test]
        fn $name() {
            use crate::parser::Parser;

            let contents = include_str!(concat!("../../testdata/input/", $path));
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path($out);
            settings.bind(|| {
                insta::assert_snapshot!(contents
                    .lines()
                    .filter_map(|line| if line != "" {
                        Some(format!(
                            "{line}\n{}",
                            $func(&mut Parser::new(line.parse().unwrap()))
                        ))
                    } else {
                        None
                    })
                    .collect::<Vec<String>>()
                    .join("\n\n"));
            });
        }
    };
}

pub(crate) use error;
#[allow(unused_imports)]
pub(crate) use snapshot;

use self::file::File;

#[derive(Debug)]
pub struct ParserError {
    stack: Vec<ParserErrorStack>,
    err: String,
}

impl std::fmt::Display for ParserError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:", self.err)?;
        for stack in &self.stack {
            write!(f, "{stack}")?;
        }

        Ok(())
    }
}

#[derive(Debug)]
pub(crate) struct ParserErrorStack {
    name: &'static str,
    #[allow(dead_code)]
    file: &'static str,
    location: (u32, u32),
}

impl std::fmt::Display for ParserErrorStack {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}({};{})", self.name, self.location.0, self.location.0)
    }
}

#[derive(Debug)]
pub struct Parser {
    tokens: VecDeque<Token>,
}

impl Parser {
    pub fn new(tokens: Tokens) -> Self {
        Self {
            tokens: VecDeque::from(tokens.0),
        }
    }

    fn first(&self) -> Option<&Token> {
        if self.tokens.len() > 0 {
            Some(self.tokens.index(0))
        } else {
            None
        }
    }

    fn nth(&mut self, nth: usize) -> Option<&Token> {
        if self.tokens.len() > nth {
            Some(self.tokens.index(nth))
        } else {
            None
        }
    }

    fn pop_front_err(&mut self, func: &'static str) -> Result<Token, ParserError> {
        self.tokens
            .pop_front()
            .ok_or(error!(func, "Expected more tokens".to_string()))
    }

    fn pop_front(&mut self) -> Option<Token> {
        self.tokens.pop_front()
    }
}

pub(crate) fn from_file(path: &PathBuf) -> Result<String, ParserError> {
    File::try_from(&mut Parser::new(
        std::fs::read_to_string(&path)
            .map_err(|err| {
                error!(
                    "from_file",
                    format!("Got an err while opening the file at {path:#?}: {err:#?}")
                )
            })?
            .parse()
            .map_err(|err| {
                error!(
                    "from_file",
                    format!("Something went wrong while tokenizing file at {path:#?}: {err:#?}")
                )
            })?,
    ))
    .map(|f| f.to_string())
}
