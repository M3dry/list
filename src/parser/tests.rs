#[cfg(test)]
mod tests {
    use crate::parser::{
        defun::Defun, exp::Exp, lambda::Lambda, r#enum::Enum, r#if::If, r#let::Let, r#match::Match,
        r#struct::Struct, r#use::Use,
    };

    macro_rules! snapshot {
        ($name:tt, $func:expr, $path:tt) => {
            snapshot!(
                $name,
                |parser| format!("{:#?}", $func(parser)),
                $path,
                "../../testdata/parser/"
            );
        };
        ($name:tt, $func:expr, $path:tt, rust) => {
            snapshot!(
                $name,
                |parser| match $func(parser) {
                    Ok(res) => res.to_string(),
                    Err(err) => format!("{err:#?}"),
                },
                $path,
                "../../testdata/rust/"
            );
        };
        ($name:tt, $func:expr, $path:tt, $out:literal) => {
            #[test]
            fn $name() {
                use crate::parser::Parser;

                let contents = include_str!(concat!("../../testdata/input/", $path));
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path($out);
                settings.bind(|| {
                    insta::assert_snapshot!(contents
                        .lines()
                        .filter_map(|line| if line != "" {
                            Some(format!(
                                "{line}\n{}",
                                $func(&mut Parser::new(line.parse().unwrap()))
                            ))
                        } else {
                            None
                        })
                        .collect::<Vec<String>>()
                        .join("\n\n"));
                });
            }
        };
    }

    snapshot!(test_calling, Exp::try_from, "calling.lt");
    snapshot!(test_calling_rust, Exp::try_from, "calling.lt", rust);
    snapshot!(test_defun, Defun::try_from, "defun.lt");
    snapshot!(test_defun_rust, Defun::try_from, "defun.lt", rust);
    snapshot!(test_if, If::try_from, "if.lt");
    snapshot!(test_if_rust, If::try_from, "if.lt", rust);
    snapshot!(test_match, Match::try_from, "match.lt");
    snapshot!(test_match_rust, Match::try_from, "match.lt", rust);
    snapshot!(test_struct, Struct::try_from, "struct.lt");
    snapshot!(test_struct_rust, Struct::try_from, "struct.lt", rust);
    snapshot!(test_enum, Enum::try_from, "enum.lt");
    snapshot!(test_enum_rust, Enum::try_from, "enum.lt", rust);
    snapshot!(test_let, Let::try_from, "let.lt");
    snapshot!(test_let_rust, Let::try_from, "let.lt", rust);
    snapshot!(test_lambda, Lambda::try_from, "lambda.lt");
    snapshot!(test_lambda_rust, Lambda::try_from, "lambda.lt", rust);
    snapshot!(test_use, Use::try_from, "use.lt");
    snapshot!(test_use_rust, Use::try_from, "use.lt", rust);
}
