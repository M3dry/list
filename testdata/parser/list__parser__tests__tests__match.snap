---
source: src/parser/tests.rs
expression: "contents.lines().filter_map(|line|\n                if line != \"\" {\n                        Some(format!(\"{line}\\n{}\",\n                                |parser| format!(\"{:#?}\", Match::try_from (parser))\n                                (& mut Parser :: new(line.parse().unwrap()))))\n                    } else { None }).collect::<Vec<String>>().join(\"\\n\\n\")"
---
(match 100 (10 1) (_ 0))
Ok(
    Match {
        against: Literal(
            Int(
                Int(
                    false,
                    100,
                ),
            ),
        ),
        branches: [
            Branch {
                pattern: Literal(
                    Int(
                        Int(
                            false,
                            10,
                        ),
                    ),
                ),
                check: None,
                ret: Literal(
                    Int(
                        Int(
                            false,
                            1,
                        ),
                    ),
                ),
            },
            Branch {
                pattern: Variable(
                    "_",
                ),
                check: None,
                ret: Literal(
                    Int(
                        Int(
                            false,
                            0,
                        ),
                    ),
                ),
            },
        ],
    },
)

(match (Option->None) (Option->None 1) ((Option->Some var) if (= var 10) 0) ((Option->Some _) 3))
Ok(
    Match {
        against: TypeCreation(
            Vars(
                Space(
                    Identifier(
                        "Option",
                    ),
                    Str(
                        Identifier(
                            "None",
                        ),
                    ),
                ),
                [],
            ),
        ),
        branches: [
            Branch {
                pattern: Enum(
                    Space(
                        Identifier(
                            "Option",
                        ),
                        Str(
                            Identifier(
                                "None",
                            ),
                        ),
                    ),
                ),
                check: None,
                ret: Literal(
                    Int(
                        Int(
                            false,
                            1,
                        ),
                    ),
                ),
            },
            Branch {
                pattern: EnumVars(
                    Space(
                        Identifier(
                            "Option",
                        ),
                        Str(
                            Identifier(
                                "Some",
                            ),
                        ),
                    ),
                    [
                        Variable(
                            "var",
                        ),
                    ],
                ),
                check: Some(
                    Infix(
                        Equality(
                            Variable(
                                "var",
                            ),
                            Literal(
                                Int(
                                    Int(
                                        false,
                                        10,
                                    ),
                                ),
                            ),
                        ),
                    ),
                ),
                ret: Literal(
                    Int(
                        Int(
                            false,
                            0,
                        ),
                    ),
                ),
            },
            Branch {
                pattern: EnumVars(
                    Space(
                        Identifier(
                            "Option",
                        ),
                        Str(
                            Identifier(
                                "Some",
                            ),
                        ),
                    ),
                    [
                        Variable(
                            "_",
                        ),
                    ],
                ),
                check: None,
                ret: Literal(
                    Int(
                        Int(
                            false,
                            3,
                        ),
                    ),
                ),
            },
        ],
    },
)

(match (StructType { x->100 y->"hello" }) ((StructType { x->100 y->"hi" }) 3) ((StructType { x->var y->"hello" }) var) (_ 0))
Ok(
    Match {
        against: TypeCreation(
            Struct(
                Str(
                    Identifier(
                        "StructType",
                    ),
                ),
                [
                    (
                        "x",
                        Literal(
                            Int(
                                Int(
                                    false,
                                    100,
                                ),
                            ),
                        ),
                    ),
                    (
                        "y",
                        Literal(
                            String(
                                "hello",
                            ),
                        ),
                    ),
                ],
            ),
        ),
        branches: [
            Branch {
                pattern: Struct(
                    Str(
                        Identifier(
                            "StructType",
                        ),
                    ),
                    [
                        Left(
                            (
                                "x",
                                Literal(
                                    Int(
                                        Int(
                                            false,
                                            100,
                                        ),
                                    ),
                                ),
                            ),
                        ),
                        Left(
                            (
                                "y",
                                Literal(
                                    String(
                                        "hi",
                                    ),
                                ),
                            ),
                        ),
                    ],
                ),
                check: None,
                ret: Literal(
                    Int(
                        Int(
                            false,
                            3,
                        ),
                    ),
                ),
            },
            Branch {
                pattern: Struct(
                    Str(
                        Identifier(
                            "StructType",
                        ),
                    ),
                    [
                        Left(
                            (
                                "x",
                                Variable(
                                    "var",
                                ),
                            ),
                        ),
                        Left(
                            (
                                "y",
                                Literal(
                                    String(
                                        "hello",
                                    ),
                                ),
                            ),
                        ),
                    ],
                ),
                check: None,
                ret: Variable(
                    "var",
                ),
            },
            Branch {
                pattern: Variable(
                    "_",
                ),
                check: None,
                ret: Literal(
                    Int(
                        Int(
                            false,
                            0,
                        ),
                    ),
                ),
            },
        ],
    },
)
